### PR TITLE
ngx.thread.kill - give all threads the permissions to kill each other

### PR DESCRIPTION
**Pull request by aviramc**
Removed the restriction that only a parent of a thread can kill it, making the ngx.thread.kill API a bit more like the UNIX-style interface. In my code, I have a case in which a thread should kill another (non child) thread.

### CODE REVIEW

**Comment by agentzh** [comment]
@aviramc Care must be take in the case when, for example, the parent is currently waiting on the thread being killed by a 3rd party. Will you add some test cases for such corner cases? I usually dare not merge anything that does not come with tests. Thanks!

**Comment by aviramc** [comment]
@agentzh I will tests, but this may take some time. If you can review the changes beforehand, that would be great. This applies to all my open pull requests (there'll be more really soon)

**Comment by agentzh** [comment]
@aviramc Yeah, sure. Thank you :)

**Comment by agentzh** [comment]
@aviramc When you kill a thread being waited by its parent, I think you should notify the sleeping parent somehow.

**Comment by aviramc** [comment]
@agentzh when you say a sleeping parent, do you mean a parent that is waiting on the thread?

**Comment by agentzh** [comment]
@aviramc Yes, via `ngx.thread.wait()`.

**Comment by aviramc** [comment]
@agentzh , please review commit  4ca58fe. In order to notify the parent thread that it the child thread has been killed, I've used posted `ngx_http_lua_post_thread()`. In order to be able to pass parameters to the posted thread, I've added the nrets attribute to the posted thread object. If you approve this fix, I'll starting writing some test cases for the feature.

File: src/ngx_http_lua_uthread.c [hunk 1]
```diff
@@ -273,6 +273,15 @@ ngx_http_lua_uthread_kill(lua_State *L)
         ngx_http_lua_del_thread(r, L, ctx, sub_coctx);
         ctx->uthreads--;
 
+        /* notify the parent that the thread was killed */
+        parent_coctx = sub_coctx->parent_co_ctx;
+        if (sub_coctx->waited_by_parent && parent_coctx) {
+            parent_co = parent_coctx->co;
+            lua_pushboolean(parent_co, 0);
+            lua_pushstring(parent_co, "user thread killed");
+            ngx_http_lua_post_thread(r, ctx, parent_coctx, 2);
```
Comment (agentzh) [inline src/ngx_http_lua_uthread.c]: Hmm, I'm a bit worried about potential race conditions here. Is it possible that the parent coroutine has already been posted at this point?

Comment (aviramc) [inline src/ngx_http_lua_uthread.c]: @agentzh , currently, parents are posted only when their children die naturally. However, in order to protect such future cases, we can add a check in `ngx_http_lua_post_thread` itself, so that a thread won't be posted twice anyway (I don't see why that would be useful), or to check that here. What do you think?

Comment (agentzh) [inline src/ngx_http_lua_uthread.c]: @aviramc When a light thread kills several siblings of it in series (without yielding in between) and their parent thread wait on all of them at the same time, then the parent thread will be posted here for multiple times, right? We need a test case for this corner case.

File: src/ngx_http_lua_uthread.c [hunk 2]
```diff
@@ -273,6 +273,15 @@ ngx_http_lua_uthread_kill(lua_State *L)
         ngx_http_lua_del_thread(r, L, ctx, sub_coctx);
         ctx->uthreads--;
 
+        /* notify the parent that the thread was killed */
+        parent_coctx = sub_coctx->parent_co_ctx;
+        if (sub_coctx->waited_by_parent && parent_coctx) {
```
Comment (agentzh) [inline src/ngx_http_lua_uthread.c]: @aviramc I think there is a corner case that a child light thread kills its parent and the parent is waiting on this child. We need to clear the `sub_coctx->waited_by_parent` field here at least, right? Since nobody is waiting on this killer thread any more. Thoughts?

File: src/ngx_http_lua_uthread.c (hunk 1 above)
Comment (agentzh) [inline src/ngx_http_lua_uthread.c]: @aviramc Also, maybe we should avoid the overhead of posting the parent thread if the killer is indeed the parent.

File: t/127-uthread-kill.t [hunk 3]
```diff
@@ -507,3 +507,112 @@ thread created: zombie
...
+               ngx.sleep(1)
+               ngx.thread.kill(thread_to_kill)
+               ngx.say("Killing thread ended")
+            end
+            
+            function sleepy_thread()
+               local i
+               for i = 1, 3 do
+                  ngx.say("Sleepy thread going to sleep")
+                  ngx.sleep(3)
```
Comment (agentzh) [inline t/127-uthread-kill.t]: @aviramc Can we use smaller sleeping intervals so that the tests can run much faster? Tests are meant to run over and over again, making them as fast as possible can be a big win :) I think times like `0.3` and `0.2` (seconds) should be good enough.

File: t/127-uthread-kill.t [hunk 4]
```diff
@@ -507,3 +507,112 @@ thread created: zombie
 lua tcp socket abort resolver
 --- error_log
 
+=== TEST 10: non parent kill, parent waits for the killed
+--- config
+    location = /t {
+        content_by_lua '
+            function killing_thread(thread_to_kill)
+               ngx.sleep(1)
```
Comment (agentzh) [inline t/127-uthread-kill.t]: @aviramc Can we use smaller sleeping intervals so that the tests can run much faster? Tests are meant to run over and over again, making them as fast as possible can be a big win :) I think times like 0.1 and 0.2 (seconds) should be long enough for our purposes.

File: t/127-uthread-kill.t [hunk 5]
```diff
@@ -507,3 +507,112 @@ thread created: zombie
...
+[error]
+[alert]
+
+
+=== TEST 11: non parent kill, parent waits for the killer
+--- config
+    location = /t {
+        content_by_lua '
+            function killing_thread(thread_to_kill)
+               ngx.sleep(3)
```
Comment (agentzh) [inline t/127-uthread-kill.t]: Let's use short sleeping times. 3 seconds in a single case are way too long for unit tests.

File: src/ngx_http_lua_uthread.c [hunk 6]
```diff
@@ -242,9 +242,9 @@ ngx_http_lua_uthread_kill(lua_State *L)
         return 2;
     }
 
-    if (sub_coctx->parent_co_ctx != coctx) {
+    if (ctx->cur_co_ctx == sub_coctx) {
         lua_pushnil(L);
-        lua_pushliteral(L, "killer not parent");
+        lua_pushliteral(L, "not allowed to kill self");
```
Comment (agentzh) [inline src/ngx_http_lua_uthread.c]: "not allowed to commit suicide" sounds better to me :)

**Comment by agentzh** [comment]
@aviramc Sorry for the long delay on my side! I've made some more comments. Will you please look into them? Thanks a lot!

File: t/127-uthread-kill.t [hunk 7]
```diff
@@ -507,3 +507,112 @@ thread created: zombie
...
+Sleepy thread going to sleep
+Parent now waits
+Killing thread ended
+Result: false user thread killed
+Finished
+
+--- no_error_log
+[error]
+[alert]
+
```
Comment (agentzh) [inline t/127-uthread-kill.t]: @aviramc Will you please run the [reindex](https://github.com/agentzh/old-openresty/blob/master/bin/reindex) tool against the `.t` files you edited to ensure the formatting is good? Thanks!

**Comment by Roycohen** [comment]
@agentzh can there be a progress on that matter.
It will be great if Openresty will close automatically all threads.
